File: 0069-sqrtx/0069-sqrtx.cpp
class Solution {
public:
    int mySqrt(int n) {
//         if(x ==0 ) return 0;
//         if(x < 4) return 1;
//         long long s = 0;
//         long long e  = x;
        
//         while(s <=e)
//         {
//             long long mid = s+ (e-s)/2;
//             if(mid == x/mid) return mid;
//             else if( mid* mid < x) s = mid+1;
//                 else e = mid-1;
//         }
//         return s-1;
        if(n ==0) return 0;
         long long s = 1;
    long long e = n;
    int ans = 1;

    while(s<=e)
    {
        long long m = s + (e-s)/2;
        // if(m*m == n) return m;

        if(m*m > n)
        {
            e = m-1;
        }
        else {
            ans = m;
            s= m+1;
        }
    }
    return ans;
    }
};

File: 0069-sqrtx/0069-sqrtx.java
class Solution {
    public int mySqrt(int x) {
        int s = 0; // start should be 0
    int e = x;
    int ans = 0;

    if (x == 0) return 0;  // base case for x = 0
    
    while (s <= e) {
        int mid = s + (e - s) / 2;
        long midSquared = (long)mid * mid; // to avoid overflow

        if (midSquared <= x) {
            ans = mid;  // mid is a possible answer
            s = mid + 1;  // search in the right half
        } else {
            e = mid - 1;  // search in the left half
        }
    }

    return ans;
    }
}